src/model/layer/convolution.cc

```diff
@@ -97,24 +97,136 @@ void Convolution::Setup(const Shape& in_sample, const LayerConf &conf) {
 
 /// \copydoc Layer::Forward(int flag, const Tensor&)
 const Tensor Convolution::Forward(int flag, const Tensor &input) {
-  Tensor output;
-  // will be used in cpp version later
+  CHECK(buf_.empty());
+  CHECK_EQ(input.device()->lang(), kCpp);
+  CHECK_EQ(input.nDim(), 4u);
+  if (flag & kTrain) buf_.push(input);
+  size_t batchsize = input.shape(0);
+  size_t imagesize = input.Size() / batchsize;
+  DataType dtype = input.data_type();
+  auto dev = input.device();
+  Shape shape{batchsize, num_filters_, conv_height_, conv_width_};
+  Tensor output(shape, dev, dtype);
   Tensor col_data(Shape{col_height_, col_width_});
-  Tensor col_grad(Shape{col_height_, col_width_});
+  float *data_col = new float[col_height_ * col_width_];
+  auto in_data = input.data<float>();
+  for (size_t b = 0; b < batchsize; b++) {
```

Comment:
according to http://arxiv.org/abs/1606.04487, if we do im2col for all images inside one mini-batch and then do gemm, the performance would be much better.
may apply such optimization in the future.